src/InstrumentationEngine/CorProfilerInfoWrapper.cpp

```diff
@@ -442,32 +442,74 @@ HRESULT MicrosoftInstrumentationEngine::CCorProfilerInfoWrapper::SetILFunctionBo
     CComPtr<CAppDomainCollection> pAppDomainCollection;
     IfFailRet(m_pProfilerManager->GetAppDomainCollection((IAppDomainCollection**)&pAppDomainCollection));
 
+    bool userBufferEnabled = InstrumentationEngineFeatures::UserBufferEnabled();
+
     CComPtr<CModuleInfo> pModuleInfo;
-    if (SUCCEEDED(pAppDomainCollection->GetModuleInfoById(moduleId, (IModuleInfo**)&pModuleInfo)))
+    if (SUCCEEDED(hr = pAppDomainCollection->GetModuleInfoById(moduleId, (IModuleInfo**)&pModuleInfo)))
     {
         CComPtr<CMethodInfo> pMethodInfo;
         hr = pModuleInfo->GetMethodInfoByToken(methodToken, &pMethodInfo);
+
         if (SUCCEEDED(hr))
         {
-            if (m_pMethodMalloc == nullptr)
+            if (!userBufferEnabled)
             {
-                CLogging::LogError(_T("Incorrect buffer passed to SetFunctionBodyMalloc"));
-                return E_FAIL;
+                if (m_pMethodMalloc == nullptr)
+                {
+                    CLogging::LogError(_T("Incorrect buffer passed to SetFunctionBodyMalloc"));
+                    return E_FAIL;
+                }
+                BYTE* pBuffer;
+                ULONG cbBuffer = 0;
+                IfFailRet(m_pMethodMalloc->GetCurrentBufferAndLen(&pBuffer, &cbBuffer));
+
+                if (pbNewILMethodHeader != pBuffer)
+                {
+                    CLogging::LogError(_T("Incorrect buffer passed to SetFunctionBodyMalloc"));
+                    return E_FAIL;
+                }
+
+                // NOTE: Relying on the size of the buffer that was handed out by the IMethodMalloc.
+                // this relies on the serialization of jit events by the profiler manager.
+                // Calculating the actual method size would require a fourth parse of header.
```

Comment:
>fourth [](http://example.com/codeflow?start=70&length=6)

force?

Reply:
Fourth was intended. We parse the header several times but don't need to here since we know the size of the buffer that was allocated.